src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -119,21 +155,47 @@ async Task EnsureLong()
 
                 try
                 {
+                    var timeoutCt = CancellationToken.None;
+                    if (async && Timeout > TimeSpan.Zero)
+                    {
+                        // We reuse the timeout's cancellation token source as long as it hasn't fired, but once it has
+                        // there's no way to reset it (see https://github.com/dotnet/runtime/issues/4694)
+                        _timeoutCts.CancelAfter(Timeout);
+                        if (_timeoutCts.IsCancellationRequested)
+                        {
+                            _timeoutCts.Dispose();
+                            _timeoutCts = new CancellationTokenSource(Timeout);
+                        }
+                        timeoutCt = _timeoutCts.Token;
+                    }
+
                     var totalRead = 0;
                     while (count > 0)
                     {
                         var toRead = Size - FilledBytes;
                         var read = async
-                            ? await Underlying.ReadAsync(Buffer, FilledBytes, toRead)
+                            ? await Underlying.ReadAsync(Buffer, FilledBytes, toRead, timeoutCt)
                             : Underlying.Read(Buffer, FilledBytes, toRead);
 
                         if (read == 0)
                             throw new EndOfStreamException();
                         count -= read;
                         FilledBytes += read;
                         totalRead += read;
+
+                        // Most of the time, it should be fine to reset cancellation token source, so we can use it again
+                        // It's still possible for cancellation token to cancel between reading and resetting (although highly improbable)
+                        // In this case, we consider it as timed out and fail with OperationCancelledException on next ReadAsync
+                        // Or we consider it not timed out if we have already read everything (count == 0)
+                        // In which case we reinitialize it on the next call to EnsureLong()
+                        if (async)
+                            _timeoutCts.CancelAfter(Timeout);
                     }
 
+                    // Resetting cancellation token source, so we can use it again
+                    if (async)
```

Comment:
It generally feels like this unified sync/async method branches a lot on async (i.e. lots of stuff specific only to one path but not the other). This may be an indication that it should be refactored into two separate methods (which possibly use some common logic in a separate inlined method).

But let's not hold this PR up on that, we can look at refactoring later.